packages/grid/x-data-grid/src/hooks/features/columnMenu/useGridColumnMenu.ts

```diff
@@ -47,6 +46,14 @@ export const useGridColumnMenu = (
   );
 
   const hideColumnMenu = React.useCallback<GridColumnMenuApi['hideColumnMenu']>(() => {
+    const field = apiRef.current.state.columnMenu.field;
+    const columnLookup = apiRef.current.state.columns.lookup;
+    // next visible column to fallback to if target column gets removed
+    const fallbackField = apiRef.current.getVisibleColumns()[0]?.field || field;
+    if (field) {
+      apiRef.current.setColumnHeaderFocus(columnLookup[field] ? field : fallbackField);
```

Comment:
Should this (fallback) logic be handled internally in `setColumnHeaderFocus` 🤔

Reply:
This one is interesting, there's also a similar problem when clicking the "Hide column".
Ideally, we should focus the closest column to the one that was removed/hidden.
I've experimented with it a bit and ended up with something like this:
```diff
diff --git a/packages/grid/x-data-grid/src/components/menu/columnMenu/menuItems/GridColumnMenuHideItem.tsx b/packages/grid/x-data-grid/src/components/menu/columnMenu/menuItems/GridColumnMenuHideItem.tsx
index 31aae86..e663b14 100644
--- a/packages/grid/x-data-grid/src/components/menu/columnMenu/menuItems/GridColumnMenuHideItem.tsx
+++ b/packages/grid/x-data-grid/src/components/menu/columnMenu/menuItems/GridColumnMenuHideItem.tsx
@@ -29,11 +29,8 @@ function GridColumnMenuHideItem(props: GridColumnMenuItemProps) {
       if (disabled) {
         return;
       }
+      apiRef.current.setColumnVisibility(colDef.field, false);
       onClick(event);
-      // time for the transition
-      timeoutRef.current = setTimeout(() => {
-        apiRef.current.setColumnVisibility(colDef.field, false);
-      }, 100);
     },
     [apiRef, colDef.field, onClick, disabled],
   );
diff --git a/packages/grid/x-data-grid/src/hooks/features/columnMenu/useGridColumnMenu.ts b/packages/grid/x-data-grid/src/hooks/features/columnMenu/useGridColumnMenu.ts
index 11fe26f..5aa5b38 100644
--- a/packages/grid/x-data-grid/src/hooks/features/columnMenu/useGridColumnMenu.ts
+++ b/packages/grid/x-data-grid/src/hooks/features/columnMenu/useGridColumnMenu.ts
@@ -4,6 +4,11 @@ import { useGridLogger, useGridApiMethod, useGridApiEventHandler } from '../../u
 import { gridColumnMenuSelector } from './columnMenuSelector';
 import { GridColumnMenuApi } from '../../../models';
 import { GridStateInitializer } from '../../utils/useGridInitializeState';
+import {
+  gridColumnLookupSelector,
+  gridColumnVisibilityModelSelector,
+  gridColumnFieldsSelector,
+} from '../columns/gridColumnsSelector';
 
 export const columnMenuStateInitializer: GridStateInitializer = (state) => ({
   ...state,
@@ -46,12 +51,32 @@ export const useGridColumnMenu = (
   );
 
   const hideColumnMenu = React.useCallback<GridColumnMenuApi['hideColumnMenu']>(() => {
-    const field = apiRef.current.state.columnMenu.field;
-    const columnLookup = apiRef.current.state.columns.lookup;
-    // next visible column to fallback to if target column gets removed
-    const fallbackField = (apiRef.current.getVisibleColumns()[0]?.field || field) as string;
-    if (field) {
-      apiRef.current.setColumnHeaderFocus(columnLookup[field] ? field : fallbackField);
+    const columnMenuState = gridColumnMenuSelector(apiRef.current.state);
+
+    if (columnMenuState.field) {
+      const columnLookup = gridColumnLookupSelector(apiRef);
+      const columnVisibilityModel = gridColumnVisibilityModelSelector(apiRef);
+      const orderedFields = gridColumnFieldsSelector(apiRef);
+      let fieldToFocus = columnMenuState.field;
+
+      // If the column was removed from the grid, we need to find the closest visible field
+      if (!columnLookup[fieldToFocus]) {
+        fieldToFocus = orderedFields[0];
+      }
+
+      // If the field to focus is hidden, we need to find the closest visible field
+      if (columnVisibilityModel[fieldToFocus] === false) {
+        // contains visible column fields + the field that was just hidden
+        const visibleOrderedFields = orderedFields.filter((field) => {
+          if (field === fieldToFocus) {
+            return true;
+          }
+          return columnVisibilityModel[field] !== false;
+        });
+        const fieldIndex = visibleOrderedFields.indexOf(fieldToFocus);
+        fieldToFocus = visibleOrderedFields[fieldIndex + 1] || visibleOrderedFields[fieldIndex - 1];
+      }
+      apiRef.current.setColumnHeaderFocus(fieldToFocus);
     }
 
     const shouldUpdate = apiRef.current.setState((state) => {

```

Reply:
Thanks for coming up with this solution, adding to this, I feel using `gridVisibleColumnFieldsSelector` may simplify it a bit

```diff
diff --git a/packages/grid/x-data-grid/src/hooks/features/columnMenu/useGridColumnMenu.ts 
b/packages/grid/x-data-grid/src/hooks/features/columnMenu/useGridColumnMenu.ts
index 11fe26f7d..f98a7f5a4 100644
--- a/packages/grid/x-data-grid/src/hooks/features/columnMenu/useGridColumnMenu.ts
+++ b/packages/grid/x-data-grid/src/hooks/features/columnMenu/useGridColumnMenu.ts
@@ -4,6 +4,7 @@ import { useGridLogger, useGridApiMethod, useGridApiEventHandler } from '.
./../u
 import { gridColumnMenuSelector } from './columnMenuSelector';
 import { GridColumnMenuApi } from '../../../models';
 import { GridStateInitializer } from '../../utils/useGridInitializeState';
+import { gridVisibleColumnFieldsSelector } from '../columns';
 
 export const columnMenuStateInitializer: GridStateInitializer = (state) => ({
   ...state,
@@ -46,12 +47,18 @@ export const useGridColumnMenu = (
   );
 
   const hideColumnMenu = React.useCallback<GridColumnMenuApi['hideColumnMenu']>(() => {
-    const field = apiRef.current.state.columnMenu.field;
-    const columnLookup = apiRef.current.state.columns.lookup;
-    // next visible column to fallback to if target column gets removed
-    const fallbackField = (apiRef.current.getVisibleColumns()[0]?.field || field) as string;
-    if (field) {
-      apiRef.current.setColumnHeaderFocus(columnLookup[field] ? field : fallbackField);
+    const columnMenuState = gridColumnMenuSelector(apiRef.current.state);
+
+    if (columnMenuState.field) {
+      const visibleColumnFields = gridVisibleColumnFieldsSelector(apiRef);
+      let fieldToFocus = columnMenuState.field;
+
+      // If the column was removed from the grid, we need to default to the first visible field
+      if (visibleColumnFields.length > 0 && !visibleColumnFields.includes(fieldToFocus)) 
{
+        fieldToFocus = visibleColumnFields[0];
+      }
+
+      apiRef.current.setColumnHeaderFocus(fieldToFocus);
     }
 
     const shouldUpdate = apiRef.current.setState((state) => {
```

Reply:
The same problem exists on `onClickAway` too, which is not working with even with this solution for some mysterious reason although `apiRef.current.setColumnHeaderFocus(fieldToFocus)` is getting called with the correct field. 🙈 

By the way, I am wondering what should be the user's expectations regarding focus when clicking outside of Grid, technically it's `<body />` the user is clicking on, so the focus going back to `<body />` would be a right thing in that case?
CC @joserodolfofreitas

Reply:
> Thanks for coming up with this solution, adding to this, I feel using gridVisibleColumnFieldsSelector may simplify it a bit

It is simpler indeed, but there's a small issue with it - the focus always goes to the first column if the target column is hidden. This doesn't work well with column virtualization when the first column might not even be rendered.
This is the reason I tried to find the closest visible column to the one that was hidden:

https://user-images.githubusercontent.com/13808724/208928612-237bbfd7-81ae-4932-ab63-094542ed3811.mov

To be clear, this is out of this PR's scope and we can address this in a separate PR 🙂